hpy/devel/include/common/hpytype.h

```diff
@@ -53,10 +53,7 @@ typedef struct {
 
 #define HPy_TPFLAGS_DEFAULT (_Py_TPFLAGS_HEAPTYPE | _Py_TPFLAGS_HAVE_VERSION_TAG)
 
-#define HPy_CAST(ctx, return_type, h) ((return_type *) _HPy_Cast(ctx, h))
-
-// XXX: These are dummy implementations and will be replaced before merging!
-#define HPy_CastPure(ctx, h) (_HPy_Cast(ctx, h) + sizeof(struct {HPyObject_HEAD}))
+#define HPy_Cast(ctx, h) (_HPy_Cast(ctx, h) + sizeof(struct {HPyObject_HEAD}))
 #define HPy_CastLegacy(ctx, h) (_HPy_Cast(ctx, h))
 
```

Comment:
I think that now that `Cast` and `CastLegacy` are two different things, there is no longer the need to use macros around `_HPy_Cast`. I propose to:
1. kill `_HPy_Cast`
2. introduce `HPy_Cast` and `HPy_CastLegacy` as two normal API functions. This will allow the debug mode to do runtime checks to ensure that you don't call the wrong cast on the wrong type
3. I tend to agree with @rlamy that `Cast` is a bit of the wrong name now, since it's no longer doing a real cast (contrarily to the old `HPy_CAST`, which did a C-level cast). Your choice whether you want to use a better name already in this PR, or if you want to merge it as is and discuss a new name later, but in that case please open an issue so that we don't forget :)

Reply:
_HPy_Cast removed and ordinary API functions added!

Let's leave renaming for another branch. I will make an issue.

Reply:
Renaming ended up happening in this branch anyway. :)